rfcs/066-graph_pipeline/README.md

```diff
@@ -0,0 +1,244 @@
+  `MERGE` instead of `CREATE`. This will ensure that repeated runs of the pipeline on the same source data will be
+  idempotent without requiring additional pipeline logic for tracking which items have already been inserted into the
+  graph.
+
+## Architecture
+
+The catalogue graph pipeline will run on a serverless event-driven architecture. All processing will be done by two
+Lambda functions, an `extractor` Lambda function, and an `indexer` Lambda function.
+
+The `extractor` will be responsible for:
+
+1. Communicating with concept sources and extracting a stream of raw source concepts via `Source` classes.
+2. Transforming each item in the stream into a source concept via `Transformer` classes, utilising the corresponding
+   Pydantic models.
+3. Batching multiple items and building final openCypher queries using the `CypherQueryBuilder` before sending the
+   queries into an SNS topic, which will be connected to an SQS queue.
+
+The `indexer` will be responsible for:
+
+1. Consuming queries in the SQS queue via
+   an [event source mapping](https://docs.aws.amazon.com/lambda/latest/dg/invocation-eventsourcemapping.html) and
+   executing the queries against the cluster using
+   the `NeptuneClient` class. Importantly, the `indexer` will be agnostic to the content of the queries, all logic
+   constructing the queries will be contained in the `extractor`.
+
+There are several reasons for separating out the `indexer` functionality into its own Lambda function and placing it
+behind an SQS queue:
+
+1. Due to the large number of indexed items (hundreds of thousands to millions), indexing from a single Lambda would
+   likely exceed the maximum allowed execution time of 15 minutes. (We experimentally verified that we can index at most
+   100 items per second to the Neptune cluster from a single thread, even when utilising batching via openCypher's
+   UNWIND clause.)
+2. Consuming from an SQS queue will allow multiple functions to run in parallel, allowing us to index items more quickly
+   without having to implement multithreading logic in our code. (Overwhelming the cluster with multiple running Lambdas
+   is not a concern as the serverless cluster should automatically scale based on traffic.)
+3. The event source mapping can be configured to insert a query back into the queue in case of error, providing an
+   automatic solution to transient faults. Additionally, queries which repeatedly fail will be placed in the DLQ, where
+   they can be investigated.
+
+The `extractor` could also be separated into several smaller microservices, but this would add a cost and performance
+overhead, as processed items would have to be serialised and deserialised between individual steps.
+
+Each invocation of the `extractor` Lambda function will only execute one Transformer, based on the input passed to the
+Lambda function. Execution will be orchestrated via AWS Step Functions. Execution order will follow the ordering in the
```

Comment:
Step-functions interactions (IIIRC) lend themselves to being diagrammed nicely, which might be a useful thing to add here.